src/main/java/org/rumbledb/compiler/VariableDependenciesVisitor.java

```diff
@@ -0,0 +1,466 @@
+
+    private RumbleRuntimeConfiguration rumbleRuntimeConfiguration;
+
+    Map<Node, Set<String>> inputVariableDependencies;
+    Map<Node, Set<String>> outputVariableDependencies;
+
+    private void addInputVariableDependencies(Node node, Set<String> variables) {
+        if (variables == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.inputVariableDependencies.keySet().contains(node)) {
+            this.inputVariableDependencies.put(node, new TreeSet<String>());
+        }
+        this.inputVariableDependencies.get(node).addAll(variables);
+    }
+
+    private void removeInputVariableDependencies(Node node, Set<String> variables) {
+        if (variables == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.inputVariableDependencies.keySet().contains(node)) {
+            return;
+        }
+        for (String v : variables) {
+            this.inputVariableDependencies.get(node).remove(v);
+        }
+    }
+
+    private void addInputVariableDependency(Node node, String variable) {
+        if (variable == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.inputVariableDependencies.keySet().contains(node)) {
+            this.inputVariableDependencies.put(node, new TreeSet<String>());
+        }
+        this.inputVariableDependencies.get(node).add(variable);
+    }
+
+    private void removeInputVariableDependency(Node node, String variable) {
+        if (variable == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.inputVariableDependencies.keySet().contains(node)) {
+            return;
+        }
+        this.inputVariableDependencies.get(node).remove(variable);
+    }
+
+    private void addOutputVariableDependencies(Node node, Set<String> variables) {
+        if (variables == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.outputVariableDependencies.keySet().contains(node)) {
+            this.outputVariableDependencies.put(node, new TreeSet<String>());
+        }
+        this.outputVariableDependencies.get(node).addAll(variables);
+    }
+
+    private void addOutputVariableDependency(Node node, String variable) {
+        if (variable == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.outputVariableDependencies.keySet().contains(node)) {
+            this.outputVariableDependencies.put(node, new TreeSet<String>());
+        }
+        this.outputVariableDependencies.get(node).add(variable);
+    }
+
+
+
+    VariableDependenciesVisitor(RumbleRuntimeConfiguration rumbleRuntimeConfiguration) {
```

Comment:
Can we move the constructor to the top of the file for house-keeping?

Reply:
DONE